### PR TITLE
[full ci] allow ssh and shell enable via vic-machine

### LINKED ISSUE #1218: Appliance SSH status and password
Should we, or should we not, leave SSH installed and enabled on the appliance for TP3? If we do it should be a backup to the vicadmin log collection otherwise we will not improve that where it's lacking.

Irrespective, the root password for the appliance needs to be altered and we need to document how to determine what the password is - standard seems to be randomly generate and echo on the VM console.

**TP3 update:**
SSH is disabled by default on the appliance. To re-enable:
1. log into the appliance VM console using root/password
2. run `systemctl start sshd`

This will persist until appliance is rebooted.

### PR DESCRIPTION
**Pull request by hickeng**
This disables console and ssh access to the appliance VM by default.
Access can be restored by running vic-machine debug, with password and/or
ssh authorized key

This is something of a kludge on the vic-machine side at the moment as I
fully expect it to be merged into `vic-machine configure` when that's available.

Fixes #1218

### CODE REVIEW

File: cmd/vic-init/toolbox.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/toolbox"
+)
+
+// startCommand is the switch for the synthetic commands that are permitted within the appliance.
+// This is not intended to allow arbitrary commands to be executed.
+func startCommand(r *toolbox.VixMsgStartProgramRequest) (int, error) {
+	switch r.ProgramPath {
```
Comment (fdawg4l) [inline cmd/vic-init/toolbox.go]: May want to log that there's a command being processed here.

File: cmd/vic-init/toolbox.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		err := fmt.Errorf("Failed to launch chsh: %s", err)
+		log.Error(err)
+		return err
+	}
+
+	// ignore the error - it's likely raced with child reaper, we just want to make sure
+	// that it's exited by the time we pass this point
+	chsh.Wait()
+
+	// confirm the change
```
Comment (fdawg4l) [inline cmd/vic-init/toolbox.go]: Is this really required?  It's not enough to check for non-negative exit status from `chsh`?

Comment (fdawg4l) [inline cmd/vic-init/toolbox.go]: Ok, now I read something about a child reaper below and I think I understand; you can't rely on exit status.  I get it.   Disregard my comment.

File: lib/install/management/debug.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+func (d *Dispatcher) DebugVCH(vch *vm.VirtualMachine, conf *config.VirtualContainerHostConfigSpec, password, authorized_key string) error {
+	defer trace.End(trace.Begin(conf.Name))
+
+	op, err := trace.FromContext(d.ctx)
+	if err != nil {
+		op = trace.NewOperation(d.ctx, "enable appliance debug")
```
Comment (fdawg4l) [inline lib/install/management/debug.go]: \m/

File: cmd/vic-init/toolbox.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+		log.Error(err)
+		return err
+	}
+
+	_, err = stdin.Write([]byte("root:" + pass))
+
+	// so that we're actively waiting when the process exits, or we'll race (and lose) to child reaper
+	var waitErr error
+	go func() {
+		waitErr = setPasswd.Wait()
```
Comment (fdawg4l) [inline cmd/vic-init/toolbox.go]: Doesn't look like `waitErr` is needed.  How about just `_ = setPasswd.Wait()`?

File: cmd/vic-init/toolbox.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/toolbox"
+)
+
+// startCommand is the switch for the synthetic commands that are permitted within the appliance.
+// This is not intended to allow arbitrary commands to be executed.
+func startCommand(r *toolbox.VixMsgStartProgramRequest) (int, error) {
+	switch r.ProgramPath {
+	case "enable-ssh":
+		return -1, enableSSH(r.Arguments)
```
Comment (fdawg4l) [inline cmd/vic-init/toolbox.go]: Not really a comment.  Looking at `VixCommandHandler`, it's not immediately obvious what the `int` is used for.  Can you clarify the `-1`?

File: cmd/vic-machine/debug/debug.go [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	executor *management.Dispatcher
+
+	enableSSH     bool
+	password      string
+	authorizedKey string
+}
+
+func NewDebug() *Debug {
+	d := &Debug{}
+	d.Data = data.NewData()
```
Comment (fdawg4l) [inline cmd/vic-machine/debug/debug.go]: We have the best package names.

**Comment by fdawg4l** [comment]
:shipit:

File: cmd/vic-init/toolbox.go (hunk 5 above)
Comment (hickeng) [inline cmd/vic-init/toolbox.go]: @dougm help me out here - this was borrowed from you code.

File: cmd/vic-init/toolbox.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+			err := fmt.Errorf("unable to create authorized_keys: %s", err)
+			log.Error(err)
+			return err
+		}
+	}
+
+	// because init is explicitly reaping child processes we cannot use simple
+	// exec commands to gather status
+	startSSH := exec.Command("/usr/bin/systemctl", "start", "sshd")
+	out, err := startSSH.CombinedOutput()
```
Comment (dougm) [inline cmd/vic-init/toolbox.go]: looks like you'd want to check and/or return err here

**Comment by dougm** [comment]
lgtm

Keep in mind wrt StartProgramInGuest, RPC calls are serialized on the VMX side as well as within the toolbox.   If we're doing anything in a start program impl that may take longer than a few hundred millis, we may want to consider making it async.  We'd kick off a process and/or go routine and return a pid (real or pseudo) before the work is done, to prevent blocking the RPC queue.  Downside is the client then needs to poll for status, such as with GuestProcessManager.ListProcessesInGuest and of course we'd need to maintain some sort of state table within the toolbox.  Doesn't look like we need to at this point, just wanted to make a note of it.

File: cmd/vic-init/toolbox.go (hunk 1 above)
Comment (hickeng) [inline cmd/vic-init/toolbox.go]: done

File: cmd/vic-machine/debug/debug.go (hunk 6 above)
Comment (hickeng) [inline cmd/vic-machine/debug/debug.go]: and so much stuttering

File: cmd/vic-init/toolbox.go (hunk 7 above)
Comment (hickeng) [inline cmd/vic-init/toolbox.go]: Can't - there's no guarantee that Wait gets the exit value for the process given the child reaper. We just have to hope (or attempt to confirm as with the prior operations)

Comment (dougm) [inline cmd/vic-init/toolbox.go]: Let's not wait for Wait here then, how about:
```go
func startSSH() error {
	c := exec.Command("/usr/bin/systemctl", "start", "sshd")

	var b bytes.Buffer
	c.Stdout = &b
	c.Stderr = &b

	if err := c.Start(); err != nil {
		return err
	}

	go func() {
		_ = c.Wait()
		log.Info("Attempted to start ssh service:\n %s", b.String())
	}()

	return nil
}
```

File: cmd/vic-init/toolbox.go [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	log.Info("Attempted to start ssh service:\n %s", out)
+
+	return nil
```
Comment (dougm) [inline cmd/vic-init/toolbox.go]: then `return startSSH()` here

**Comment by hickeng** [comment]
poke

**Comment by hickeng** [comment]
poke

**Pull request merged**